### PR TITLE
Michael

### PR DESCRIPTION
**Pull request by MichaelRes**
developments single particle reconstruction

### CODE REVIEW

File: notebooks/Create or Load Density Maps.ipynb [hunk 1]
```diff
@@ -42,14 +42,21 @@
...
-   "execution_count": 10,
+   "execution_count": 4,
    "metadata": {},
    "outputs": [
     {
-     "name": "stdout",
-     "output_type": "stream",
-     "text": [
-      "> Loaded map of shape: (272, 272, 272)\n"
+     "ename": "FileNotFoundError",
```
Comment (fredericpoitevin) [inline notebooks/Create or Load Density Maps.ipynb:49]: download of those maps did not work?

File: notebooks/Create or Load Density Maps.ipynb [hunk 2]
```diff
@@ -122,6 +129,126 @@
...
+    "#to visualize the 3d map\n",
+    "from mpl_toolkits.mplot3d import Axes3D\n",
+    "from matplotlib import cm\n",
+    "IMG_DIM = 32\n",
+    "\n",
+    "def normalize(arr):\n",
+    "    arr_min = np.min(arr)\n",
+    "    return (arr-arr_min)/(np.max(arr)-arr_min)\n",
+    "\n",
+    "def explode(data):\n",
```
Comment (fredericpoitevin) [inline notebooks/Create or Load Density Maps.ipynb:148]: I am not sure I understand why you need to do this?

File: notebooks/Generating a dataset.ipynb [hunk 3]
```diff
@@ -67,12 +68,33 @@
     "    \n",
     "    def show(self):\n",
     "        if self.has_image:\n",
-    "            plt.imshow(self.image, cmap='Greys')            "
+    "            plt.imshow(self.image, cmap='Greys') \n",
+    "            \n",
+    "    def create_map(self):\n",
```
Comment (fredericpoitevin) [inline notebooks/Generating a dataset.ipynb:73]: we should make this routine more efficient / smarter by only updating the voxels within a given radius from an atom.

File: notebooks/Generating a dataset.ipynb [hunk 4]
```diff
@@ -144,21 +150,50 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
    "metadata": {},
    "outputs": [],
    "source": [
-    "dataset = np.load('dataset 3 atoms (1000 images).npy')"
+    "dataset = np.load('generated_dataset/dataset 3 atoms (1000 images).npy')"
```
Comment (fredericpoitevin) [inline notebooks/Generating a dataset.ipynb:176]: going forward, I suggest we do not push data to the github repo.

**Review by fredericpoitevin — APPROVED** [review]
I will check the new Single particle reconstruction notebook asap.

See other comments, but good job!

You can merge once you saw the comments

File: notebooks/Create or Load Density Maps.ipynb (hunk 1 above)
Comment (MichaelRes) [reply notebooks/Create or Load Density Maps.ipynb:49]: I get this error: FileNotFoundError: [Errno 2] No such file or directory: '../data/EMPIAR/10077/emd_4121.map.gz'

File: notebooks/Create or Load Density Maps.ipynb (hunk 2 above)
Comment (MichaelRes) [reply notebooks/Create or Load Density Maps.ipynb:148]: This cell is actually just a copy pasta of the functions in the link above, I didn't dig into it ! But the plot these function give are better than the ones with basic plt functions so I used them

File: notebooks/Generating a dataset.ipynb (hunk 3 above)
Comment (MichaelRes) [reply notebooks/Generating a dataset.ipynb:73]: yes that sounds doable, I'll check that

File: notebooks/Generating a dataset.ipynb (hunk 4 above)
Comment (MichaelRes) [reply notebooks/Generating a dataset.ipynb:176]: I might need help on how to be able to remotely download the data without having to have it on the repo...

File: notebooks/Create or Load Density Maps.ipynb (hunk 1 above)
Comment (fredericpoitevin) [reply notebooks/Create or Load Density Maps.ipynb:49]: didn't you use the notebook https://github.com/fredericpoitevin/gmm-cryoem/blob/master/notebooks/Download%20maps%20from%20EMDB.ipynb ?

File: notebooks/Create or Load Density Maps.ipynb (hunk 2 above)
Comment (fredericpoitevin) [reply notebooks/Create or Load Density Maps.ipynb:148]: OK, I don't think it's very useful but if it helps you: good!

File: notebooks/Generating a dataset.ipynb (hunk 4 above)
Comment (fredericpoitevin) [reply notebooks/Generating a dataset.ipynb:176]: oh no, the idea would be that you generate it locally, but don't add it to the files that you push to github. With the code I can also generate locally. This way we keep the repo light and tidy ;)
The .gitignore file is here for that

Comment (MichaelRes) [reply notebooks/Generating a dataset.ipynb:176]: oh ok, I get it now !

File: notebooks/Create or Load Density Maps.ipynb [hunk 5]
```diff
@@ -289,17 +283,101 @@
...
+    "    return volume\n",
+    "\n",
+    "def init_volume_bis(n, L, crds=None, rads=None):\n",
+    "    \"\"\"\n",
+    "    \"\"\"\n",
+    "    ctr = 0\n",
+    "    volume = np.zeros((n,n,n))\n",
+    "\n",
+    "    for iat in range(len(crds)):\n",
+    "        #coordinate of the voxel that contains the center of the particle\n",
```
Comment (fredericpoitevin) [inline notebooks/Create or Load Density Maps.ipynb:297]: yes! that's the idea! ;)

File: notebooks/Single particle reconstruction.ipynb [hunk 6]
```diff
@@ -147,7 +207,81 @@
...
     "\n",
-    "plt.imshow(np.absolute(im_))"
+    "plt.imshow(np.real(im_))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Si j'essaie avec la slice à x=0"
```
Comment (fredericpoitevin) [inline notebooks/Single particle reconstruction.ipynb:217]: je vais regarder les notebooks asap

**Review by fredericpoitevin — APPROVED** [review]

File: notebooks/Create or Load Density Maps.ipynb (hunk 1 above)
Comment (MichaelRes) [reply notebooks/Create or Load Density Maps.ipynb:49]: It works now !

**Pull request merged**